File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/GRPC/ProjectsGRPCService.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// GRPC/ProjectsGRPCService.swift
//
// gRPC service stub for projects operations.
// =============================================================================

import Foundation
import GRPC

// MARK: - Response Models

/// Project response data.
public struct ProjectResponse: Sendable {
    public let id: String
    public let name: String
    public let description: String?
    public let ownerId: String
    public let createdAt: Int64
    public let updatedAt: Int64
    
    public init(
        id: String,
        name: String,
        description: String?,
        ownerId: String,
        createdAt: Int64,
        updatedAt: Int64
    ) {
        self.id = id
        self.name = name
        self.description = description
        self.ownerId = ownerId
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }
}

/// Pagination response data.
public struct PaginationResponse: Sendable {
    public let page: Int
    public let pageSize: Int
    public let totalPages: Int
    public let totalItems: Int
    
    public init(page: Int, pageSize: Int, totalPages: Int, totalItems: Int) {
        self.page = page
        self.pageSize = pageSize
        self.totalPages = totalPages
        self.totalItems = totalItems
    }
}

/// Projects list response.
public struct ProjectsListResponse: Sendable {
    public let projects: [ProjectResponse]
    public let pagination: PaginationResponse
    
    public init(projects: [ProjectResponse], pagination: PaginationResponse) {
        self.projects = projects
        self.pagination = pagination
    }
}

// MARK: - Service

/// gRPC service stub for projects operations.
/// Wraps the Grey Projects gRPC service.
public final class ProjectsGRPCService: Sendable {
    
    private let channel: GreyChannel
    
    // MARK: - Initialization
    
    public init(channel: GreyChannel) {
        self.channel = channel
    }
    
    // MARK: - Operations
    
    /// List projects with pagination.
    /// - Parameters:
    ///   - accessToken: Current access token
    ///   - page: Page number (1-indexed)
    ///   - pageSize: Items per page
    /// - Returns: ProjectsListResponse on success
    /// - Throws: GreyError on failure
    public func listProjects(
        accessToken: String,
        page: Int = 1,
        pageSize: Int = 20
    ) async throws -> ProjectsListResponse {
        do {
            let grpcChannel = channel.getChannel()
            let options = channel.callOptions(accessToken: accessToken)
            
            // Note: In production, call actual gRPC stub
            // let client = Projects_ProjectsServiceAsyncClient(channel: grpcChannel)
            // let request = Projects_ListProjectsRequest.with {
            //     $0.page = Int32(page)
            //     $0.pageSize = Int32(pageSize)
            // }
            // let response = try await client.listProjects(request, callOptions: options)
            
            return ProjectsListResponse(
                projects: [],
                pagination: PaginationResponse(
                    page: page,
                    pageSize: pageSize,
                    totalPages: 0,
                    totalItems: 0
                )
            )
        } catch let error as GRPCStatus {
            throw GreyError.fromGRPCStatus(error)
        } catch {
            throw GreyError.fromError(error)
        }
    }
    
    /// Create a new project.
    /// - Parameters:
    ///   - accessToken: Current access token
    ///   - name: Project name
    ///   - description: Optional project description
    /// - Returns: Created ProjectResponse on success
    /// - Throws: GreyError on failure
    public func createProject(
        accessToken: String,
        name: String,
        description: String?
    ) async throws -> ProjectResponse {
        do {
            let grpcChannel = channel.getChannel()
            let options = channel.callOptions(accessToken: accessToken)
            
            // Note: In production, call actual gRPC stub
            // let client = Projects_ProjectsServiceAsyncClient(channel: grpcChannel)
            // let request = Projects_CreateProjectRequest.with {
            //     $0.name = name
            //     $0.description_p = description ?? ""
            // }
            // let response = try await client.createProject(request, callOptions: options)
            
            let now = Int64(Date().timeIntervalSince1970 * 1000)
            return ProjectResponse(
                id: "",
                name: name,
                description: description,
                ownerId: "",
                createdAt: now,
                updatedAt: now
            )
        } catch let error as GRPCStatus {
            throw GreyError.fromGRPCStatus(error)
        } catch {
            throw GreyError.fromError(error)
        }
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/grpc/ProjectsGrpcService.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// grpc/ProjectsGrpcService.kt
//
// gRPC service stub for projects operations.
// =============================================================================

package com.grey.sdk.grpc

import com.grey.sdk.config.GreyOptions
import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import io.grpc.StatusException
import io.grpc.StatusRuntimeException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Project data model.
 */
data class ProjectResponse(
    val id: String,
    val name: String,
    val description: String?,
    val ownerId: String,
    val createdAt: Long,
    val updatedAt: Long
)

/**
 * Pagination data model.
 */
data class PaginationResponse(
    val page: Int,
    val pageSize: Int,
    val totalPages: Int,
    val totalItems: Int
)

/**
 * Projects list response.
 */
data class ProjectsListResponse(
    val projects: List<ProjectResponse>,
    val pagination: PaginationResponse
)

/**
 * Create project request.
 */
data class CreateProjectRequest(
    val name: String,
    val description: String?
)

/**
 * gRPC service stub for projects operations.
 * Wraps the Grey Projects gRPC service.
 */
class ProjectsGrpcService(
    private val channel: GreyChannel,
    private val options: GreyOptions
) {
    /**
     * List projects with pagination.
     *
     * @param accessToken Current access token
     * @param page Page number (1-indexed)
     * @param pageSize Number of items per page
     * @return GreyResult containing ProjectsListResponse or error
     */
    suspend fun listProjects(
        accessToken: String,
        page: Int = 1,
        pageSize: Int = 20
    ): GreyResult<ProjectsListResponse> {
        return withContext(Dispatchers.IO) {
            try {
                val grpcChannel = channel.getChannel()
                
                // Note: In production, call actual gRPC stub with auth metadata
                // val metadata = channel.createAuthMetadata(accessToken)
                // val stub = ProjectsServiceGrpcKt.ProjectsServiceCoroutineStub(grpcChannel)
                //     .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
                // val response = stub.listProjects(listProjectsRequest {
                //     this.page = page
                //     this.pageSize = pageSize
                // })
                
                // Simulated response - replace with actual gRPC call
                GreyResult.success(
                    ProjectsListResponse(
                        projects = emptyList(),
                        pagination = PaginationResponse(
                            page = page,
                            pageSize = pageSize,
                            totalPages = 0,
                            totalItems = 0
                        )
                    )
                )
            } catch (e: StatusException) {
                GreyResult.failure(GreyError.fromGrpcStatus(e))
            } catch (e: StatusRuntimeException) {
                GreyResult.failure(GreyError.fromGrpcStatusRuntime(e))
            } catch (e: Throwable) {
                GreyResult.failure(GreyError.fromException(e))
            }
        }
    }

    /**
     * Create a new project.
     *
     * @param accessToken Current access token
     * @param name Project name
     * @param description Optional project description
     * @return GreyResult containing created Project or error
     */
    suspend fun createProject(
        accessToken: String,
        name: String,
        description: String?
    ): GreyResult<ProjectResponse> {
        return withContext(Dispatchers.IO) {
            try {
                val grpcChannel = channel.getChannel()
                
                // Note: In production, call actual gRPC stub with auth metadata
                // val metadata = channel.createAuthMetadata(accessToken)
                // val stub = ProjectsServiceGrpcKt.ProjectsServiceCoroutineStub(grpcChannel)
                //     .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
                // val response = stub.createProject(createProjectRequest {
                //     this.name = name
                //     this.description = description ?: ""
                // })
                
                // Simulated response - replace with actual gRPC call
                GreyResult.success(
                    ProjectResponse(
                        id = "", // From gRPC response
                        name = name,
                        description = description,
                        ownerId = "",
                        createdAt = System.currentTimeMillis(),
                        updatedAt = System.currentTimeMillis()
                    )
                )
            } catch (e: StatusException) {
                GreyResult.failure(GreyError.fromGrpcStatus(e))
            } catch (e: StatusRuntimeException) {
                GreyResult.failure(GreyError.fromGrpcStatusRuntime(e))
            } catch (e: Throwable) {
                GreyResult.failure(GreyError.fromException(e))
            }
        }
    }
}
